packages/code-infra/src/utils/extractErrorCodes.mjs

```diff
@@ -0,0 +1,163 @@
+ */
+
+/**
+ * Extracts error codes from all files in a directory.
+ * @param {import('../cli/pnpm.mjs').PublicPackage} pkg
+ * @param {Set<string>} errors
+ * @param {import('@mui/internal-babel-plugin-minify-errors').Options['detection']} [detection='opt-in']
+ */
+async function extractErrorCodesForPackage(pkg, errors, detection = 'opt-in') {
+  const srcPath = path.join(pkg.path, 'src');
+  const srcPathExists = await fs
+    .stat(srcPath)
+    .then((stat) => stat.isDirectory())
+    .catch(() => false);
+  // Implementation to extract error codes from all files in the directory
+  const cwd = srcPathExists ? srcPath : pkg.path;
+  const files = await globby('**/*.{js,ts,jsx,tsx,cjs,mjs,cts}', {
+    ignore: [
+      '**/node_modules/**',
+      '**/dist/**',
+      '**/build/**',
+      'scripts',
+      '**/__{tests,fixtures,mock,mocks}__/**',
+      ...BASE_IGNORES,
+    ],
+    cwd,
+  });
+  console.log(`🔍 ${pkg.name}: Processing ${files.length} file${files.length > 1 ? 's' : ''}...`);
+  await wrapInWorker(
+    async (file) => {
+      const fullPath = path.join(cwd, file);
+      const code = await fs.readFile(fullPath, 'utf8');
+      const ast = await parseAsync(code, {
+        filename: fullPath,
+        sourceType: 'module',
+        plugins: [[babelSyntaxTypescript, { isTSX: true }], [babelSyntaxJsx]],
+        configFile: false,
+        babelrc: false,
+        browserslistConfigFile: false,
+        code: false,
+      });
+      if (!ast) {
+        throw new Error(`Failed to parse ${fullPath}`);
+      }
+      traverse(ast, {
+        NewExpression(newExpressionPath) {
+          const { message } =
+            findMessageNode(babelTypes, newExpressionPath, {
+              detection,
+              missingError: 'annotate',
+            }) ?? {};
+          if (message) {
+            errors.add(message.message);
+          }
+        },
+      });
+    },
+    { defaultConcurrency: 30, items: files },
+  );
+}
+
+/**
+ * Extracts error codes from all workspace packages.
+ * @param {Args} args
+ */
+export default async function extractErrorCodes(args) {
+  const { skip: skipPackages = [], errorCodesPath, detection = 'opt-in' } = args;
+  /**
+   * @type {Set<string>}
+   */
+  const errors = new Set();
+
+  const basePackages = await getWorkspacePackages({
+    publicOnly: true,
+  });
+  /**
+   * @type {Set<string>}
+   */
+  const skippedPackages = new Set();
+  const packages = basePackages.filter((pkg) =>
+    // Ignore obvious packages that do not have user-facing errors
+    {
+      const shouldNotSkip =
+        !pkg.name.startsWith('@mui/internal-') &&
+        !pkg.name.startsWith('@mui-internal/') &&
+        !skipPackages.includes(pkg.name);
+      if (!shouldNotSkip) {
+        skippedPackages.add(pkg.name);
+      }
+      return shouldNotSkip;
+    },
+  );
+  await Promise.all(packages.map((pkg) => extractErrorCodesForPackage(pkg, errors, detection)));
```

Comment:
Would it make sense to determine the full list of files across all packages, and then run the extractor concurrently over that. Otherwise we multiply the concurrency with each specified package. i.e. 1 package => concurrency 30, 3 packages => concurrency becomes 90.